### ISSUE TITLE
Ctrl+C does not work in PTY's

### ISSUE THREAD

**Issue by thecodingwizard** [google/gvisor#9898]
### Description

At Modal, we ran into an issue where Ctrl+C does not work consistently when a binary creates a new PTY.

Specifically, I created a script `shell.c` that:

1. Creates a PTY with `forkpty()`
2. In the child process, runs `/bin/bash`
3. In the parent process, pipes stdin/stdout to the pty's master file descriptor

When running `shell.c`, Ctrl+C stops working after the first command has finished executing in `/bin/bash`.

I do not observe this issue when running `shell.c` with `runc`.

### Steps to reproduce

I have reproduced the issue both with Docker and with `runsc exec`.

#### With Docker

```
docker run --rm -it --runtime=runsc ubuntu:latest /bin/bash
apt update
apt install vim build-essential -y
```

Create a file `shell.c`:

<details>
<summary>shell.c</summary>

```c
// ctrl+c only works the first time

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <pty.h>
#include <utmp.h>
#include <sys/select.h>
#include <termios.h>
#include <unistd.h>

int main() {
    pid_t pid;
    int master_fd;

    close(0);

    // Fork and create a new PTY
    pid = forkpty(&master_fd, NULL, NULL, NULL);

    if (pid < 0) {
        perror("forkpty");
        exit(EXIT_FAILURE);
    }

    if (pid == 0) { // Child process
        // Execute /bin/bash in the child process
        execl("/bin/bash", "bash", (char *) NULL);
        perror("execl");
        exit(EXIT_FAILURE);
    } else { // Parent process

        // give /bin/bash enough time to spawn
        usleep(500 * 1000);

        write(master_fd, "\x03", 1);
        usleep(100 * 1000); // give /bin/bash enough time to finish processing the input
        write(master_fd, "whoami\n", 7);
        usleep(100 * 1000);
        write(master_fd, "\x03", 1);
        usleep(100 * 1000);
        write(master_fd, "whoami\n", 7);
        usleep(100 * 1000);
        write(master_fd, "\x04", 1);


        char buffer[16384];
        ssize_t bytes_read = read(master_fd, buffer, sizeof(buffer));
        write(STDOUT_FILENO, buffer, bytes_read);
    }

    return 0;
}
```
</details>

```
gcc -o shell shell.c -lutil
./shell
```

Expected output:

```
root@2e676828fec9:/# ^C
root@2e676828fec9:/# whoami
root
root@2e676828fec9:/# ^C
root@2e676828fec9:/# whoami
root
root@2e676828fec9:/#
```

Actual output:

```
root@337c88183991:~# whoami
root
root@337c88183991:~# whoami
root
root@337c88183991:~#
```

With docker, even the first ^C does not appear. With `runsc exec`, the first ^C works.

If you remove the `--runtime=runsc` flag in the `docker run` command, the correct behavior is observed.

#### With `runsc exec`

In your container, create a file `/root/shell.c` (same as above):

<details>
<summary>/root/shell.c</summary>

```c
// ctrl+c only works the first time

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <pty.h>
#include <utmp.h>
#include <sys/select.h>
#include <termios.h>
#include <unistd.h>

int main() {
    pid_t pid;
    int master_fd;

    close(0);

    // Fork and create a new PTY
    pid = forkpty(&master_fd, NULL, NULL, NULL);

    if (pid < 0) {
        perror("forkpty");
        exit(EXIT_FAILURE);
    }

    if (pid == 0) { // Child process
        // Execute /bin/bash in the child process
        execl("/bin/bash", "bash", (char *) NULL);
        perror("execl");
        exit(EXIT_FAILURE);
    } else { // Parent process

        // give /bin/bash enough time to spawn
        usleep(500 * 1000);

        write(master_fd, "\x03", 1);
        usleep(100 * 1000); // give /bin/bash enough time to finish processing the input
        write(master_fd, "whoami\n", 7);
        usleep(100 * 1000);
        write(master_fd, "\x03", 1);
        usleep(100 * 1000);
        write(master_fd, "whoami\n", 7);
        usleep(100 * 1000);
        write(master_fd, "\x04", 1);


        char buffer[16384];
        ssize_t bytes_read = read(master_fd, buffer, sizeof(buffer));
        write(STDOUT_FILENO, buffer, bytes_read);
    }

    return 0;
}
```
</details>

Compile with `gcc -o shell shell.c -lutil`. Then, run `runsc exec [container-id] /root/shell`. The expected result is to see something like this:

```
root@modal:~# ^C
root@modal:~# whoami
root
root@modal:~# ^C
root@modal:~# whoami
root
root@modal:~#
```

Instead, I see:

```
root@modal:~# ^C
root@modal:~# whoami
root
root@modal:~# whoami
root
root@modal:~#
```

Note that the second `^C` is missing. The correct behavior is observed when using `runc`.

The full use case is to spawn a fully-interactive `/bin/bash` with a new PTY, piping stdin/stdout to/from the master PTY.

<details>
<summary>Fully interactive `shell.c`</summary>

```c
// ctrl+c only works the first time

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <pty.h>
#include <utmp.h>
#include <sys/select.h>
#include <termios.h>

int main() {
    pid_t pid;
    int master_fd;
    struct termios orig_termios, raw;

    // Save original terminal attributes for stdin
    tcgetattr(STDIN_FILENO, &orig_termios);

    // Fork and create a new PTY
    pid = forkpty(&master_fd, NULL, NULL, NULL);

    if (pid < 0) {
        perror("forkpty");
        exit(EXIT_FAILURE);
    }

    if (pid == 0) { // Child process
        // Execute /bin/bash in the child process
        execl("/bin/bash", "bash", (char *) NULL);
        perror("execl");
        exit(EXIT_FAILURE);
    } else { // Parent process
        // Set stdin to raw mode
        raw = orig_termios;
        cfmakeraw(&raw);
        tcsetattr(STDIN_FILENO, TCSANOW, &raw);

        // Set up file descriptor sets for select
        fd_set read_fds;
        char buffer[256];
        int max_fd = master_fd + 1;

        int status;
        while (1) {
            FD_ZERO(&read_fds);
            FD_SET(STDIN_FILENO, &read_fds);
            FD_SET(master_fd, &read_fds);

            // Check if the child process has exited. Careful: buggy... can get stuck on select()
            pid_t result = waitpid(pid, &status, WNOHANG);
            if (result == pid) {
                // Child process has exited
                break;
            }

            // Wait for data on either the PTY or stdin.
            if (select(max_fd, &read_fds, NULL, NULL, NULL) < 0) {
                perror("select");
                break;
            }

            // Relay data from stdin to PTY
            if (FD_ISSET(STDIN_FILENO, &read_fds)) {
                ssize_t bytes_read = read(STDIN_FILENO, buffer, sizeof(buffer));
                if (bytes_read <= 0) {
                    break;
                }
                write(master_fd, buffer, bytes_read);
            }

            // Relay data from PTY to stdout
            if (FD_ISSET(master_fd, &read_fds)) {
                ssize_t bytes_read = read(master_fd, buffer, sizeof(buffer));
                if (bytes_read <= 0) {
                    break;
                }
                write(STDOUT_FILENO, buffer, bytes_read);
            }
        }

        // Restore original terminal attributes
        tcsetattr(STDIN_FILENO, TCSANOW, &orig_termios);

        // Wait for the child process to exit, in case it hasn't already
        waitpid(pid, &status, 0);

        // Exit after the child process has finished
        exit(WEXITSTATUS(status));
    }

    return EXIT_SUCCESS;
}
```

Compile with `gcc -o shell shell.c -lutil`. Try running `runsc exec [container-id] /root/shell`. A new bash shell should open inside the container, but ctrl+c will only work until the first command has finished executing. `runsc exec [container-id] /bin/bash` works fine.

</details>

### runsc version

```shell
runsc version 5eaa66a2ed33
spec: 1.1.0-rc.1
```

### docker version

```shell
Client: Docker Engine - Community
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.20.10
 Git commit:        afdd53b
 Built:             Thu Oct 26 09:08:01 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.10
  Git commit:       311b9ff
  Built:            Thu Oct 26 09:08:01 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.26
  GitCommit:        3dd1e886e55dd695541fdcd67420c2888645a495
 runc:
  Version:          1.1.10
  GitCommit:        v1.1.10-0-g18a0cb0
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### uname

Linux ip-10-1-5-239 5.15.0-1048-aws #53~20.04.1-Ubuntu SMP Wed Oct 4 16:44:20 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### runsc debug logs (if available)

<details>
<summary>Logs</summary>

```shell
I0117 20:21:14.151880  192333 main.go:194] **************** gVisor ****************
I0117 20:21:14.151947  192333 main.go:195] Version 5eaa66a2ed33, go1.21.1 X:nocoverageredesign, amd64, 8 CPUs, linux, PID 192333, PPID 192332, UID 0, GID 0
D0117 20:21:14.151962  192333 main.go:196] Page size: 0x1000 (4096 bytes)
I0117 20:21:14.151975  192333 main.go:197] Args: [./runsc --debug-log=/tmp/runsc_logs --debug --strace --root=/run/user/1000/runsc exec dXfaMTEaqq0 /root/shell]
I0117 20:21:14.151993  192333 config.go:391] Platform: systrap
I0117 20:21:14.152014  192333 config.go:392] RootDir: /run/user/1000/runsc
I0117 20:21:14.152020  192333 config.go:393] FileAccess: exclusive / Directfs: true / Overlay: root:self
I0117 20:21:14.152029  192333 config.go:394] Network: sandbox
I0117 20:21:14.152036  192333 config.go:396] Debug: true. Strace: true, max size: 1024, syscalls:
D0117 20:21:14.152043  192333 config.go:408] Config.RootDir (--root): /run/user/1000/runsc
D0117 20:21:14.152051  192333 config.go:408] Config.Traceback (--traceback): system
D0117 20:21:14.152059  192333 config.go:408] Config.Debug (--debug): <bool Value>
D0117 20:21:14.152066  192333 config.go:408] Config.LogFilename (--log): (empty)
D0117 20:21:14.152073  192333 config.go:408] Config.LogFormat (--log-format): text
D0117 20:21:14.152083  192333 config.go:408] Config.DebugLog (--debug-log): /tmp/runsc_logs
D0117 20:21:14.152091  192333 config.go:408] Config.DebugToUserLog (--debug-to-user-log): <bool Value>
D0117 20:21:14.152098  192333 config.go:408] Config.DebugCommand (--debug-command): (empty)
D0117 20:21:14.152105  192333 config.go:408] Config.PanicLog (--panic-log): (empty)
D0117 20:21:14.152112  192333 config.go:408] Config.CoverageReport (--coverage-report): (empty)
D0117 20:21:14.152121  192333 config.go:408] Config.DebugLogFormat (--debug-log-format): text
D0117 20:21:14.152129  192333 config.go:408] Config.FileAccess (--file-access): <config.FileAccessType Value>
D0117 20:21:14.152136  192333 config.go:408] Config.FileAccessMounts (--file-access-mounts): <config.FileAccessType Value>
D0117 20:21:14.152144  192333 config.go:408] Config.Overlay (--overlay): <bool Value>
D0117 20:21:14.152151  192333 config.go:408] Config.Overlay2 (--overlay2): <config.Overlay2 Value>
D0117 20:21:14.152158  192333 config.go:408] Config.FSGoferHostUDS (--fsgofer-host-uds): <bool Value>
D0117 20:21:14.152165  192333 config.go:408] Config.HostUDS (--host-uds): <config.HostUDS Value>
D0117 20:21:14.152173  192333 config.go:408] Config.HostFifo (--host-fifo): <config.HostFifo Value>
D0117 20:21:14.152180  192333 config.go:408] Config.Network (--network): <config.NetworkType Value>
D0117 20:21:14.152187  192333 config.go:408] Config.EnableRaw (--net-raw): <bool Value>
D0117 20:21:14.152195  192333 config.go:408] Config.AllowPacketEndpointWrite (--TESTONLY-allow-packet-endpoint-write): <bool Value>
D0117 20:21:14.152202  192333 config.go:408] Config.HostGSO (--gso): <bool Value>
D0117 20:21:14.152209  192333 config.go:408] Config.GvisorGSO (--software-gso): <bool Value>
D0117 20:21:14.152216  192333 config.go:408] Config.GvisorGROTimeout (--gvisor-gro): <time.Duration Value>
D0117 20:21:14.152223  192333 config.go:408] Config.TXChecksumOffload (--tx-checksum-offload): <bool Value>
D0117 20:21:14.152231  192333 config.go:408] Config.RXChecksumOffload (--rx-checksum-offload): <bool Value>
D0117 20:21:14.152240  192333 config.go:408] Config.QDisc (--qdisc): <config.QueueingDiscipline Value>
D0117 20:21:14.152248  192333 config.go:408] Config.LogPackets (--log-packets): <bool Value>
D0117 20:21:14.152261  192333 config.go:408] Config.PCAP (--pcap-log): (empty)
D0117 20:21:14.152268  192333 config.go:408] Config.Platform (--platform): systrap
D0117 20:21:14.152275  192333 config.go:408] Config.PlatformDevicePath (--platform_device_path): (empty)
D0117 20:21:14.152282  192333 config.go:408] Config.MetricServer (--metric-server): (empty)
D0117 20:21:14.152289  192333 config.go:408] Config.ProfilingMetrics (--profiling-metrics): (empty)
D0117 20:21:14.152296  192333 config.go:408] Config.ProfilingMetricsLog (--profiling-metrics-log): (empty)
D0117 20:21:14.152310  192333 config.go:408] Config.ProfilingMetricsRate (--profiling-metrics-rate-us): <int Value>
D0117 20:21:14.152317  192333 config.go:408] Config.Strace (--strace): <bool Value>
D0117 20:21:14.152324  192333 config.go:408] Config.StraceSyscalls (--strace-syscalls): (empty)
D0117 20:21:14.152332  192333 config.go:408] Config.StraceLogSize (--strace-log-size): <uint Value>
D0117 20:21:14.152339  192333 config.go:408] Config.StraceEvent (--strace-event): <bool Value>
D0117 20:21:14.152346  192333 config.go:410] Config.DisableSeccomp: <bool Value>
D0117 20:21:14.152353  192333 config.go:408] Config.EnableCoreTags (--enable-core-tags): <bool Value>
D0117 20:21:14.152360  192333 config.go:408] Config.WatchdogAction (--watchdog-action): <watchdog.Action Value>
D0117 20:21:14.152368  192333 config.go:408] Config.PanicSignal (--panic-signal): <int Value>
D0117 20:21:14.152408  192333 config.go:408] Config.ProfileEnable (--profile): <bool Value>
D0117 20:21:14.152417  192333 config.go:408] Config.ProfileBlock (--profile-block): (empty)
D0117 20:21:14.152424  192333 config.go:408] Config.ProfileCPU (--profile-cpu): (empty)
D0117 20:21:14.152435  192333 config.go:408] Config.ProfileHeap (--profile-heap): (empty)
D0117 20:21:14.152442  192333 config.go:408] Config.ProfileMutex (--profile-mutex): (empty)
D0117 20:21:14.152449  192333 config.go:408] Config.TraceFile (--trace): (empty)
D0117 20:21:14.152456  192333 config.go:410] Config.RestoreFile: (empty)
D0117 20:21:14.152463  192333 config.go:408] Config.NumNetworkChannels (--num-network-channels): <int Value>
D0117 20:21:14.152470  192333 config.go:408] Config.Rootless (--rootless): <bool Value>
D0117 20:21:14.152478  192333 config.go:408] Config.AlsoLogToStderr (--alsologtostderr): <bool Value>
D0117 20:21:14.152485  192333 config.go:408] Config.ReferenceLeak (--ref-leak-mode): <refs.LeakMode Value>
D0117 20:21:14.152492  192333 config.go:408] Config.CPUNumFromQuota (--cpu-num-from-quota): <bool Value>
D0117 20:21:14.152499  192333 config.go:408] Config.AllowFlagOverride (--allow-flag-override): <bool Value>
D0117 20:21:14.152506  192333 config.go:408] Config.OCISeccomp (--oci-seccomp): <bool Value>
D0117 20:21:14.152514  192333 config.go:408] Config.IgnoreCgroups (--ignore-cgroups): <bool Value>
D0117 20:21:14.152521  192333 config.go:408] Config.SystemdCgroup (--systemd-cgroup): <bool Value>
D0117 20:21:14.152528  192333 config.go:408] Config.PodInitConfig (--pod-init-config): (empty)
D0117 20:21:14.152535  192333 config.go:408] Config.BufferPooling (--buffer-pooling): <bool Value>
D0117 20:21:14.152542  192333 config.go:408] Config.XDP (--EXPERIMENTAL-xdp): <config.XDP Value>
D0117 20:21:14.152550  192333 config.go:408] Config.AFXDPUseNeedWakeup (--EXPERIMENTAL-xdp-need-wakeup): <bool Value>
D0117 20:21:14.152557  192333 config.go:408] Config.FDLimit (--fdlimit): <int Value>
D0117 20:21:14.152575  192333 config.go:408] Config.DCache (--dcache): <int Value>
D0117 20:21:14.152582  192333 config.go:408] Config.IOUring (--iouring): <bool Value>
D0117 20:21:14.152590  192333 config.go:408] Config.DirectFS (--directfs): <bool Value>
D0117 20:21:14.152597  192333 config.go:408] Config.NVProxy (--nvproxy): <bool Value>
D0117 20:21:14.152604  192333 config.go:408] Config.NVProxyDocker (--nvproxy-docker): <bool Value>
D0117 20:21:14.152611  192333 config.go:408] Config.TPUProxy (--tpuproxy): <bool Value>
D0117 20:21:14.152618  192333 config.go:408] Config.TestOnlyAllowRunAsCurrentUserWithoutChroot (--TESTONLY-unsafe-nonroot): <bool Value>
D0117 20:21:14.152625  192333 config.go:408] Config.TestOnlyTestNameEnv (--TESTONLY-test-name-env): (empty)
D0117 20:21:14.152633  192333 config.go:408] Config.TestOnlyAFSSyscallPanic (--TESTONLY-afs-syscall-panic): <bool Value>
D0117 20:21:14.152640  192333 config.go:410] Config.explicitlySet: <map[string]struct {} Value>
D0117 20:21:14.152647  192333 config.go:408] Config.ReproduceNAT (--reproduce-nat): <bool Value>
D0117 20:21:14.152654  192333 config.go:408] Config.ReproduceNftables (--reproduce-nftables): <bool Value>
I0117 20:21:14.152664  192333 main.go:199] **************** gVisor ****************
D0117 20:21:14.152706  192333 state_file.go:78] Load container, rootDir: "/run/user/1000/runsc", id: {SandboxID: ContainerID:dXfaMTEaqq0}, opts: {Exact:false SkipCheck:false TryLock:false RootContainer:false}
D0117 20:21:14.154354  192333 container.go:675] Signal container, cid: dXfaMTEaqq0, signal: signal 0 (0)
D0117 20:21:14.154388  192333 sandbox.go:1211] Signal sandbox "dXfaMTEaqq0"
D0117 20:21:14.154396  192333 sandbox.go:613] Connecting to sandbox "dXfaMTEaqq0"
D0117 20:21:14.154505  192333 urpc.go:568] urpc: successfully marshalled 91 bytes.
D0117 20:21:14.154643  192333 urpc.go:611] urpc: unmarshal success.
D0117 20:21:14.154669  192333 exec.go:129] Exec arguments: /root/shell
D0117 20:21:14.154680  192333 exec.go:130] Exec capabilities: <nil>
I0117 20:21:14.154740  192333 exec.go:148] Using exec capabilities from container: &{PermittedCaps:2818835963 InheritableCaps:2818835963 EffectiveCaps:2818835963 BoundingCaps:2818835963 AmbientCaps:0}
D0117 20:21:14.154760  192333 container.go:592] Execute in container, cid: dXfaMTEaqq0, args: /root/shell
D0117 20:21:14.154768  192333 sandbox.go:562] Executing new process in container "dXfaMTEaqq0" in sandbox "dXfaMTEaqq0"
D0117 20:21:14.154775  192333 sandbox.go:1489] Changing "/dev/stdin" ownership to 0/0
D0117 20:21:14.154788  192333 sandbox.go:1489] Changing "/dev/stdout" ownership to 0/0
D0117 20:21:14.154795  192333 sandbox.go:1489] Changing "/dev/stderr" ownership to 0/0
D0117 20:21:14.154802  192333 sandbox.go:613] Connecting to sandbox "dXfaMTEaqq0"
D0117 20:21:14.154968  192333 urpc.go:568] urpc: successfully marshalled 1479 bytes.
D0117 20:21:14.155818  192333 urpc.go:611] urpc: unmarshal success.
D0117 20:21:14.155837  192333 container.go:706] Forwarding all signals to container, cid: dXfaMTEaqq0, PIDPID: 276, fgProcess: true
D0117 20:21:14.156709  192333 container.go:663] Wait on process 276 in container, cid: dXfaMTEaqq0
D0117 20:21:14.156739  192333 sandbox.go:1165] Waiting for PID 276 in sandbox "dXfaMTEaqq0"
D0117 20:21:14.156753  192333 sandbox.go:613] Connecting to sandbox "dXfaMTEaqq0"
D0117 20:21:14.156815  192333 urpc.go:568] urpc: successfully marshalled 75 bytes.
D0117 20:21:14.968459  192333 urpc.go:611] urpc: unmarshal success.
D0117 20:21:14.968510  192333 container.go:714] Done forwarding signals to container, cid: dXfaMTEaqq0, PID: 276, fgProcess: true
I0117 20:21:14.968544  192333 main.go:226] Exiting with status: 0
```
</details>

**Comment by ayushr2** [comment]
Thanks for the detailed bug report and the reproducers! Always helpful.

Although I am not an expert in TTY/PTY codebase here, I think I have tracked down the issue and fixed it in #9912. Let me get it reviewed and I can submit it with a regression test.

**Comment by thundergolfer** [comment]
Thanks @ayushr2!
